### ISSUE TITLE
GraphHopperStorage.getEdges() includes removed edges

### ISSUE THREAD

**Issue by easbar** [graphhopper/graphhopper#2111]
After calling `GraphHopperStorage#optimize` (in `PrepareRoutingSubnetworks`) calling `getEdges()` returns a bigger number of edges than what we get when counting the edges directly like this:

```java
 int count = 0;
 AllEdgesIterator iter = ghStorage.getAllEdges();
 while (iter.next()) {
      count++;
 }

 // fails
 assertEquals(count, ghStorage.getEdges());
```

See this branch: https://github.com/graphhopper/graphhopper/tree/edge_count_bug

**Comment by otbutz** [comment]
https://github.com/graphhopper/graphhopper/blob/85f8801d6fa15284bf8acde09ac954f7464a8313/core/src/main/java/com/graphhopper/storage/GraphHopperStorage.java#L227-L229

**Comment by easbar** [comment]
Maybe the question is if `getEdges()` should return the number of edges or the maximum (+1) used edge ID. Ideally after `optimize()` we should not notice anything about removed nodes or edges and all edges are numbered consecutively.

**Comment by easbar** [comment]
Look at this test:

```java
    @Test
    public void testNodeRemoval() {
        CarFlagEncoder encoder = new CarFlagEncoder();
        EncodingManager em = EncodingManager.create(encoder);
        GraphHopperStorage graph = new GraphBuilder(em).create();
        assertEquals(0, graph.edge(0, 3, 10, true).getEdge());
        assertEquals(1, graph.edge(2, 4, 10, true).getEdge());
        assertEquals(2, graph.edge(1, 5, 10, true).getEdge());
        assertEquals(3, graph.edge(6, 0, 10, true).getEdge());
        assertEquals(7, graph.getNodes());
        assertEquals(4, graph.getEdges());
        System.out.println("graph before:");
        GHUtility.printGraphForUnitTest(graph, encoder);
// prints:
// graph.edge(0, 3, 10.000000, true); // edgeId=0
// graph.edge(2, 4, 10.000000, true); // edgeId=1
// graph.edge(1, 5, 10.000000, true); // edgeId=2
// graph.edge(6, 0, 10.000000, true); // edgeId=3

        graph.markNodeRemoved(2);
        graph.markNodeRemoved(6);
        graph.optimize();
        System.out.println("graph after:");
        GHUtility.printGraphForUnitTest(graph, encoder);
// prints:
// graph.edge(0, 3, 10.000000, true); // edgeId=0
// graph.edge(1, 2, 10.000000, true); // edgeId=2

        // todonow: fails! why is the number of nodes reduced to 5 instead of 4?
        assertEquals(4, graph.getNodes());
        // todonow: fails! why does this give 4? the number of edges is 2, and the largest edge ID is 2.
        assertEquals(2, graph.getEdges());
    }
```

Removing nodes 2 and 6 leads to the removal of the 2-4 and 6-0 edges and node 5 is relabeled to node 2, which is nice because this way the node IDs are consecutive. This is not true for the edge IDs of the remaining edges which stay the same (0 and 2). However, `graph.getNodes()` returns 5 (instead of 4?) and `graph.getEdges()` returns 4 which is neither the number of edges nor does it give a clue about the maximum edge ID.

**Comment by karussell** [comment]
> assertEquals(7, graph.getNodes());
> ...
> todonow: fails! why is the number of nodes reduced to 5 instead of 4?
> assertEquals(4, graph.getNodes());

I think because 7-2=5 :)

> This is not true for the edge IDs of the remaining edges which stay the same (0 and 2).

Yeah, this is ugly and we could probably improve this optimization code also as it has the problem with the turn costs relabelling.

> and graph.getEdges() returns 4 which is neither the number of edges

It is indeed incorrectly described in the docs and only correctly described in AllEdgesIterator.length:

> the maximum edgeId in the graph plus 1

**Comment by michaz** [comment]
I also tripped over that recently, and I _also_ distinctly remember that we've had almost this discussion before.

I agree that, no matter the _history_ of this code, that a number returned by something on AllEdgesIterator that looks like it should be a length or a size must be the number of items iterated over by AllEdgesIterator, and nothing else. Otherwise _something_ is wrong.

**Comment by michaz** [comment]
Probably resolved by #2118 if that gets merged.
